File: knapsack/knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int knapsack(vector<int> values, vector<int> weights, int capacity, int start);

int main() {
	vector<int> values = {60,100,120};
	vector<int> weights = {10,20,30};
	int capacity = 50;
	int start = 0;
	int k = knapsack(values,weights,capacity,start);
	cout << k << endl;

	return 0;
}

int knapsack(vector<int> values, vector<int> weights, int capacity, int start) {
	if (start == values.size() || capacity == 0) {
		return 0;
	}
	if (weights[start] <= capacity) {
		return max(
			values[start] + knapsack(values,weights,capacity-weights[start],start+1),
			knapsack(values,weights,capacity,start+1)
			);
	}
	else {
		return knapsack(values,weights,capacity,start+1);
	}
}

File: knapsack/knapsack.py
def knapsack(values,weights,capacity):
    if len(values) == 0 or capacity == 0:
        return 0
    if weights[0] <= capacity:
        return max(
                values[0] + knapsack(values[1:],weights[1:],capacity-weights[0]),
                knapsack(values[1:],weights[1:],capacity)
                )
    else:
        return knapsack(values[1:],weights[1:],capacity)

values = [60,100,120]
weights = [10,20,30]
capacity = 50
print(knapsack(values,weights,capacity))
print('----------')
values = [10,20,30,40]
weights = [12,13,15,19]
capacity = 10
print(knapsack(values,weights,capacity))
